Source/Core/InputCommon/ControllerInterface/WGInput/WGInput.cpp

```diff
@@ -0,0 +1,536 @@
+          g_controller_interface.AddDevice(std::move(dev));
+      }
+    }
+  }
+}
+
+void Device::PopulateButtons(Microsoft::WRL::ComPtr<WGI::IRawGameController> controller,
+                             std::vector<boolean>& buttons)
+{
+  INT32 button_count = 0;
+  controller->get_ButtonCount(&button_count);
+
+  buttons.resize(button_count);
+  u32 i = 0;
+  for (auto& button : buttons)
+  {
+    WGI::GameControllerButtonLabel lbl = WGI::GameControllerButtonLabel_None;
+    controller->GetButtonLabel(i, &lbl);
+
+    if (lbl != WGI::GameControllerButtonLabel_None && lbl < wgi_button_names.size())
+    {
+      AddInput(new NamedButton(&button, wgi_button_names[lbl]));
+    }
+    else
+    {
+      AddInput(new IndexedButton(&button, i));
+    }
+
+    ++i;
+  }
+}
+
+Gamepad::Gamepad(Microsoft::WRL::ComPtr<WGI::IRawGameController> _raw_controller,
+                 WGI::IGamepad* _gamepad, std::string _name)
+    : Device(std::move(_name)), m_raw_controller(_raw_controller), m_gamepad(_gamepad)
+{
+  // Buttons:
+  PopulateButtons(m_raw_controller, m_buttons);
+
+  // Axes:
+  const double* axis = &m_gamepad_reading.LeftThumbstickX;
+  for (auto& axis_name : gamepad_axis_names)
+  {
+    AddInput(new NamedAxis(axis, 0.0, -1.0, axis_name));
+    AddInput(new NamedAxis(axis, 0.0, +1.0, axis_name));
+    ++axis;
+  }
+
+  // Triggers:
+  const double* trigger = &m_gamepad_reading.LeftTrigger;
+  for (auto& trigger_name : gamepad_trigger_names)
+  {
+    AddInput(new NamedTrigger(trigger, trigger_name));
+    ++trigger;
+  }
+
+  // Motors:
+  double* motor = &m_state_out.LeftMotor;
+  for (auto& motor_name : gamepad_motor_names)
+  {
+    AddOutput(new NamedMotor(motor, motor_name, *this));
+    ++motor;
+  }
+
+  ResetInput();
+}
+
+RawGameController::RawGameController(
+    Microsoft::WRL::ComPtr<WGI::IRawGameController> _raw_controller, std::string _name)
+    : Device(std::move(_name)), m_raw_controller(_raw_controller)
+{
+  // Buttons:
+  PopulateButtons(m_raw_controller, m_buttons);
+
+  // Axes:
+  INT32 axis_count = 0;
+  m_raw_controller->get_AxisCount(&axis_count);
+  m_axes.resize(axis_count);
+
+  u32 i = 0;
+  for (auto& axis : m_axes)
+  {
+    // AddAnalogInputs adds additional "FullAnalogSurface" Inputs:
+    AddAnalogInputs(new IndexedAxis(&axis, 0.5, +0.5, i), new IndexedAxis(&axis, 0.5, -0.5, i));
+    ++i;
+  }
+
+  // Switches (Hats):
+  INT32 switch_count = 0;
+  m_raw_controller->get_SwitchCount(&switch_count);
+  m_switches.resize(switch_count);
+  i = 0;
+  for (auto& swtch : m_switches)
+  {
+    using gcsp = WGI::GameControllerSwitchPosition;
+
+    WGI::GameControllerSwitchKind switch_kind;
+    m_raw_controller->GetSwitchKind(i, &switch_kind);
+
+    AddInput(new Switch(&swtch, i, gcsp::GameControllerSwitchPosition_Up));
+    AddInput(new Switch(&swtch, i, gcsp::GameControllerSwitchPosition_Down));
+
+    if (WGI::GameControllerSwitchKind_TwoWay != switch_kind)
```

Comment:
Our current codebase convention doesn't utilize inverted conditionals like this (it also doesn't protect against anything here anyways).